### DISCUSSION TITLE
A few Exult Studio inquiries, please.

### DISCUSSION THREAD

**Discussion by ShadowTigerZC** [exult/exult#333] [General]
Howdy hey, all.  I just played through Glimmerscape for the nth time and it got my creative juices flowing.  So I've got a few questions, please.

1)  I remember in past editings of the map, I'd put something down, then leave and come back, and suddenly some of my object additions seem to have duplicated themselves to every copy of that particular chunk in the game.  Like in this screenshot that I may or may not have exhibited in the past.
https://i.imgur.com/w983jKk.png
So my question is, what can cause this?  I guess I should phrase the question thusly:

"If I put down a chunk, when does the game think I'm editing the chunk itself, versus when does it think I'm just putting down any other object (Which happens to be a wall or a plant, not a gold bar.) on it without editing the chunk?" 

2)  How do we actually compile the usecode?  I read a bit of the tutorial but something in it didn't quite match up to the reality of my epistemological experience, for lack of a better expression.  I see the executable, but it doesn't do anything when I hit enter on it to use it. 

3)  In Glimmerscape, if I use my inner-hacker to Pickpocket an NPC and they have a scroll on them that says "Generic Scroll" but if I actually kill them and take the scroll from their body, the scroll says something much more real.  Is that usecode that modifies that scroll, or is that just the fact that the NPC itself and the NPC's body are two completely separate things with two separate and unrelated inventories, so the scroll didn't "become" the real one, and they were just two different scrolls?  (I.e. I had this issue with the robotic Provider NPC's "code" scroll.)

4)  If I decided to add some new graphics to the game, how restricted am I to palettes in terms of my graphics editing tools?  Can I just use photoshop and import whatever I do into Exult Studio without worrying about the colors?  Or do I have to import the palettes available in Exult Studios into the graphic editing program and work on the art from there?  How straightforward would it be?

5)  If I made loose tiles available here, what would the most convenient format for people to receive it and import them straight into Exult Studios?  I.e.  "Here's a PNG with a transparent background. It's 55x65."

6)  Is there a way to set the Avatar's starting position and map?

7)  Is there a way to blank out (or delete) a map altogether with a few clicks of something?  I like to start fresh.

8)  Got any tips for rapidly creating a landmass?  :-P  Or copying and pasting chunks and not just the objects on them?

Thanks bunchly!  👋 

---------

Heck, while I'm here, here's a completely separate bunch of topics.

1)  Any thought on creating a Discord for Exult?  Having some live chat for cheap little quick responses could be useful.  It's got great drag-and-drop functionality for things like screenshots or small files, and can have any number of links and references.
--- EDIT --- I did join the Ultima Dragons Discord and I see the little area for Exult.  Howdy.

2)  Would it be useful to have an "Exult Studio" Discussion forum in addition to the others here in the Github?  Great for folks with questions like mine, above.

**Comment by DominusExult** [comment]
Some quick answers:
1. I'm not 100% sure. Mostly when you remove things that were in the original chunk it will spread.
2. http://exult.info/studio.php#usecode_edit - it's a commandline utility
3. my guess is that there is usecode that changes these scrolls in the NPC's death event
4. 255 colors indexed and sometimes bad things happen - so not a good answer
5. Best is to import these into Exult, make sure they are correct and then export as shp but also as png to easily show the shp. Especially with multiframe shapes, that's best
6. the place you set the avatar and save is the starting position. If you want to edit the original game, start Exult in mapedit mode http://exult.info/docs.php#command_line and make sure to place the avatar in the original spot when saving in ES
7. I guess manually deleting the map from the patch folder
8. command line tools mockup and smooth http://exult.info/studio.php#tools
9. 

So far the little traffic doesn't make an extra forum useful, I think

**Comment by KnightCaptainU7** [comment]
1. It may have to do with the item's Object Class in Exult Studio. "Building" and "Unusable" seem to stick to the chunk, excluding mountaintops. The logic might be listed in the Exult/ES code someplace? I think a better chunk editor would be really helpful to building new worlds or adding to existing ones.
2. The easiest way to get started is to download SIfixes and install Exult Tools. You'll need a few of the EXEs from the Tools folder for running the compile script.
3. Agree with Dominus. Event == DEATH can be used for setting global flags, etc.
4. Ultima 7 and thus Exult work by heavily relying on Indexed Color. You'll want to export the first palette from Palettes.flx to use in GIMP (2.8 or earlier) for making things. If you cycle the palette numbers you'll see how this avoids recreating every shape in a different form. For example when you are invisible nearly everything uses palette 3. This recolors everything on screen using those Indexed Colors. For a game shipped on floppy disks this space-saving was a big deal.
5. Agree. In Exult a "tile" is an 8x8 pixel square shape. I'm not sure there is a way to export chunks?
6. Agree. In SI the Avatar is standing near a proximity egg that triggers the start sequence. You could use a similar egg to transport him to the map number you want.
7. Agree.
8. Agree.
9. You can attach files here by dragging and dropping, or even pasting them. That includes images.

**Comment by KnightCaptainU7** [comment]
Exult Studio feature request: https://github.com/exult/exult/issues/336

**Comment by Donfrow** [comment]
In regards number 3 and Glimmerscape that could possibly be a bug. The NPC you mentioned was one of my earlier ones so there is not much special on it usecode wise such, such as a DEATH event. 

"Generic Unused Scroll" is the text for any undefined scroll quality number in my usecode. This includes quality 0 which all new scrolls are assigned by default from the Exult Studio when dropped into the world. This makes me think whatever method you are using to pull the scroll from the NPC without killing it is causing it to not be the "true" scroll. What method are you using to take it from the NPC? I tried it myself using Exult studio and pulling it from the gump and it was working as expected (scroll quality 59).

**Reply by ShadowTigerZC** [reply]
Ah, thanks for the response.  I figured it'd be something like that.

The method I used was indeed the "Pickpocket" method, mostly because I didn't want to kill the NPC because I didn't receive any dialogue options like "You shouldn't exist so I have to alter/kill you."   The conversation pretty much stopped with a tiny inconsequential conversation so I thought I had sequence-broke somewhere and my game was broken, so I had no choice but to get the scroll from him one way or another.  So I did.  

Weird line of thought, but there it is.

**Reply by Donfrow** [reply]
Tried using vibrate and it dropped the proper scroll as well, unless there's another way to pick pocket I'm not familiar with. Fair point about the text, I am pretty sure that was such an early NPC when I was really just messing around with things. Never wrote much of it with cohesiveness in mind.

Overall I'd say it's probably a bug in my mod or inadvertently broke something with my patchwork usecode. I definitely didn't put anything in for that behaviour specifically, though the others are correct, you could do it with the DEATH event. If you do fiddle around with something like that the DEATH event only triggers when a NPC has the tournament flag enabled.

**Reply by ShadowTigerZC** [reply]
Oh, in Exult, I think if you hit [Alt]+[P] (Or Ctrl+P?) it activates pickpocket mode, where if you double-click on an NPC, it opens their inventory.  So that's how I got the scroll, so I wouldn't have to kill that NPC and feel bad about it.

Still, again, thanks for the explanation!  Appreciated and interesting.

**Reply by KnightCaptainU7** [reply]
Why not just set the quality of the scroll the NPC is carrying? That's just a map edit.

**Reply by Donfrow** [reply]
I did not know that pick pocket mode existed, Seems it is [Alt] + [Ctrl] + P. 

Seems it works as expected being the proper scroll quality for me. At least it only happens when you cheat!

**Reply by KnightCaptainU7** [reply]
Neither did I.
